lib/postgrex/type_supervisor.ex

```diff
@@ -3,29 +3,57 @@ defmodule Postgrex.TypeSupervisor do
 
   use Supervisor
 
+  @manager Postgrex.TypeManager
+  @servers Postgrex.TypeSupervisor
+
+  @doc """
+  Starts a type supervisor with a manager and a simple
+  one for one supervisor for each server.
+  """
   def start_link(mode \\ :manager)
 
   def start_link(:manager) do
     Supervisor.start_link(__MODULE__, :manager)
   end
+
   def start_link(:servers) do
-    Supervisor.start_link(__MODULE__, :servers, [name: __MODULE__])
+    Supervisor.start_link(__MODULE__, :servers, name: @servers)
+  end
+
+  @doc """
+  Locates a type server for the given module-key pair.
+  """
+  def locate(module, key) do
+    pair = {module, key}
+
+    case Registry.lookup(@manager, pair) do
+      [{pid, _}] -> if Process.alive?(pid), do: pid, else: start_server(module, pair)
+      [] -> start_server(module, pair)
+    end
   end
 
-  def start_server(module, starter) do
-    Supervisor.start_child(__MODULE__, [module, starter])
+  defp start_server(module, pair) do
+    args = [module, self(), [name: {:via, Registry, {Postgrex.TypeManager, pair}}]]
+
+    case Supervisor.start_child(@servers, args) do
+      {:ok, pid} -> pid
+      {:error, {:already_started, pid}} -> pid
+    end
   end
 
+  # Callbacks
+
   def init(:manager) do
-    manager    = worker(Postgrex.TypeManager, [])
+    manager = worker(Registry, [:unique, @manager])
```

Comment:
Shouldn't this be `supervisor`?

Reply:
Great catch! This is why I live Elixir's child_spec, haha.